aten/src/THC/THCTensorMathReduce.cuh

```diff
@@ -426,58 +434,55 @@ __global__ void THCTensor_kernel_varInnermostDim(Real *tgt, Real *src_, unsigned
       }
     }
 
-    local_sum[threadIdx.y][threadIdx.x] =
+    Accreal local_sum =
         THCNumerics<Accreal>::mul(local_mean, ScalarConvert<int, Accreal>::to(count));
     __syncthreads();
 
-
-    // Compute the true mean of each of the blockDim.x rows by reducing the sums
-    for (unsigned s = 8; s > 1; s >>= 1) {
-      if (row < num_rows && threadIdx.x < s) {
-        local_sum[threadIdx.y][threadIdx.x] = THCNumerics<Accreal>::add(
-            local_sum[threadIdx.y][threadIdx.x],
-            local_sum[threadIdx.y][threadIdx.x + s]);
-      }
-      __syncthreads();
+    /*
+     * We are reducing across each row of 16 threads to find the true sum of the
+     * entire input row.
+     *
+     * This is done in a warp shuffle: each warp has 2 rows of 16 threads for a
+     * total of 32 threads. The warp will accumulate the sums of each row of 
+     * 16 threads in the first thread of the row (with threadIdx.x == 0)
+     */
+    for (unsigned s = 8; s >= 1; s >>= 1) {
+      local_sum = THCNumerics<Accreal>::add(local_sum, 
+          WARP_SHFL_DOWN((row < num_rows) ? local_sum : acc_zero, s));
     }
-
     if (row < num_rows && threadIdx.x == 0) {
-      mean[threadIdx.y] = THCNumerics<Accreal>::div(
-          THCNumerics<Accreal>::add(local_sum[threadIdx.y][0], local_sum[threadIdx.y][1]),
-          ScalarConvert<int, Accreal>::to(row_size));
+      // This is the true mean of the entire input row. There are 32 true means.
+      mean[threadIdx.y] = THCNumerics<Accreal>::div(local_sum, ScalarConvert<int, Accreal>::to(row_size));
```

Comment:
If I understand correctly, `mean[threadIdx.y]` is computed by lane 0 and then broadcasted to all threads in the warp. Can this be `__shfl_sync` (`WARP_SHFL` in THC)?

It's probably not much faster than the shared-memory.

Reply:
Good point, that should work. I'll update this using __shfl_sync